### PR TITLE
refactor: separate dsc into a library and a binary crate

### PR DESCRIPTION
**Pull request by EricCrosson**
The motivation for this change is so that I can import the engine of docker-source-checksum as a function in some other binary, without introducing changes to current docker-source-checksum users.

### CODE REVIEW

**Comment by EricCrosson** [comment]
Marking as "ready for review" because from what I can tell, this is working as expected! But it may be in what you consider to be a rough state, happy to iterate on this as feedback is presented

**Comment by dpc** [comment]
@EricCrosson All good, but I think underscores vs hypens are super confusing. Would putting the binary and library in the same crate (like e.g. here: https://github.com/crev-dev/cargo-crev/blob/e9b35a1d24a7b877bf20d19632e8ef9f04b210ac/cargo-crev/Cargo.toml#L17) work for you?

**Comment by ericcrosson-bitgo** [comment]
> I think underscores vs hypens are super confusing

Agreed! There may still be a bit of that, but I followed the example you linked and the diff is much cleaner. Thanks for the pointer

**Comment by EricCrosson** [comment]
> I think underscores vs hypens are super confusing

Agreed! There may still be a bit of that, but I followed the example you linked and the diff is much cleaner. Thanks for the pointer

File: src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+        .collect::<std::result::Result<Vec<_>, _>>()
+        .map_err(|e| format_err!("{}", e))?;
+
+    digests.append(
+        &mut opts
+            .extra_string
+            .iter()
+            .map(|s| s.as_bytes().to_vec())
+            .collect(),
+    );
```
Comment (ocornoc) [inline src/lib.rs]: You can avoid the unnecessary collection into a `Vec` by 
```suggestion
    digests.extend(opts.extra_string.iter().map(|s| s.as_bytes().to_vec()));
```

File: src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+
+    let mut parts = line.split_ascii_whitespace();
+    let cmd = parts.next().expect("At least the command");
+
+    match cmd {
+        "COPY" | "ADD" => {
+            let src_globs: Vec<_> = parts.map(str::to_string).collect();
+            if src_globs.is_empty() {
+                bail!("No arguments to {} command?", cmd);
+            }
```
Comment (ocornoc) [inline src/lib.rs]: Here you can use `ensure!` from the `failure` crate (which should be replaced IMO, but not in this PR. `failure` is deprecated, unmaintained, and was only a proof of concept). 
```suggestion
            failure::ensure!(!src_globs.is_empty(), "No arguments to {cmd} commands?");
```

File: src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+    for line in content
+        .lines()
+        .map(|line| line.trim())
+        .filter(|line| !line.starts_with("#"))
+    {
+        if line.ends_with("\\") {
+            prev_line.push_str(&line[..line.len() - 1]);
+        } else {
+            prev_line.push_str(line);
+            res.push(std::mem::replace(&mut prev_line, "".into()))
```
Comment (ocornoc) [inline src/lib.rs]:
```suggestion
            res.push(std::mem::take(&mut prev_line))
```

File: src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+        }
+    }
+
+    if prev_line != "" {
```
Comment (ocornoc) [inline src/lib.rs]:
```suggestion
    if !prev_line.is_empty() {
```

File: src/lib.rs (hunk 2 above)
Comment (ocornoc) [reply src/lib.rs]: I don't know the Rust version being used to compile this, but hopefully it supports string interpolation.

**Review by ocornoc — COMMENTED** [review]
I have some criticisms of this code (was requested privately by @EricCrosson).

**Comment by ericcrosson-bitgo** [comment]
Thanks for the input @ocornoc. I'd like to keep this PR as close to a lift-and-shift as possible, but I'm happy to circle back with these suggestions

**Review by dpc — APPROVED** [review]

**Comment by dpc** [comment]
@EricCrosson Proposed changes looked good so I added them. Please ping me if you're OK with my landing as is.

**Comment by EricCrosson** [comment]
@dpc Looks good to me! Thank you

**Pull request merged**